File: hw/hidalgo/u1/hw16Libraries/JavaScript/calculatorIncomeTax.js
import { finalBalance } from './incomeTaxlib.js';
import readline from 'node:readline';

class ConsoleReader {
    constructor() {
        this.io = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
    }

    ask(text) {
        return new Promise(res => this.io.question(text, res));
    }

    async getNumberFloat(label) {
        let number;
        do {
            const entry = await this.ask(label);
            number = Number(entry);
            if (isNaN(number)) {
                console.log(" Invalid number. Example: 2500.75");
            }
        } while (isNaN(number));
        return number;
    }

    async getNumberInt(label) {
        let number;
        do {
            const entry = await this.ask(label);
            number = Number(entry);
            if (!Number.isInteger(number)) {
                console.log(" Invalid integer. Example: 3");
            }
        } while (!Number.isInteger(number));
        return number;
    }

    finish() {
        this.io.close();
    }
}

class TaxSystem {
    constructor(monthSalary, personalSpend, dependents, retention) {
        this.monthSalary = monthSalary;
        this.personalSpend = personalSpend;
        this.dependents = dependents;
        this.retentionPaid = retention;
    }

    calculateOutcome() {
        const yearIncome = this.monthSalary * 12;
        const iess = yearIncome * 0.0945;
        return finalBalance(yearIncome, iess, this.personalSpend, this.dependents, this.retentionPaid);
    }

    showResult(result) {
        console.log("\n======== TAX RESULTS 2025 ========");
        result > 0
            ? console.log(`Amount to pay: $${result.toFixed(2)}`)
            : console.log(`Balance in your favor: $${Math.abs(result).toFixed(2)}`);
        console.log("==================================");
    }
}

async function startApp() {
    console.log("\n*** ECUADOR INCOME TAX CALCULATOR 2025 ***\n");

    const reader = new ConsoleReader();

    const salary = await reader.getNumberFloat("Monthly average salary: ");
    const expenses = await reader.getNumberFloat("Total personal expenses per year: ");
    const familyMembers = await reader.getNumberInt("How many dependents do you have?: ");
    const alreadyPaid = await reader.getNumberFloat("Withholdings paid in the year: ");

    reader.finish();

    const taxProcess = new TaxSystem(salary, expenses, familyMembers, alreadyPaid);
    const result = taxProcess.calculateOutcome();

    taxProcess.showResult(result);
}

startApp();

File: hw/hidalgo/u1/hw16Libraries/PythonLib/calculator_income_tax.py
import income_tax_lib as ir

def ask_number(text, whole=False):
    while True:
        try:
            return int(input(text)) if whole else float(input(text))
        except ValueError:
            print("Please enter a valid number!")

def run_calculator():
    print("\n======= INCOME TAX ECUADOR 2025 =======\n")

    monthly = ask_number("Enter monthly salary: $")
    expenses = ask_number("Enter annual personal expenses: $")
    dependents = ask_number("Number of dependents: ", True)
    withheld = ask_number("Tax already withheld: $")

    annual_income = monthly * 12
    iess = annual_income * 0.0945

    result = ir.get_net_balance(annual_income, iess, expenses, dependents, withheld)

    if result > 0:
        print(f"\n Tax to pay: ${result:.2f}")
    else:
        print(f"\n Balance in your favor: ${abs(result):.2f}")

if __name__ == "__main__":
    run_calculator()
